### PR TITLE
Add mutable mapping

### LINKED ISSUE #29: MutableMapping
It would be convenient to treat an S3 bucket as a `MutableMapping`.  

This might look something like the following `MutableMapping` around a local directory: https://github.com/mrocklin/zict/blob/master/zict/file.py

It would be useful for hosting NDArrays on S3 using an upcoming Zarr refactor: https://github.com/alimanfoo/zarr/pull/22

@martindurant does this interest you?

### PR DESCRIPTION
**Pull request by martindurant**
Can now use S3 like a dict holding bytes.

Fixes #29

### CODE REVIEW

File: s3fs/mapwrap.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    def __iter__(self):
+        return self.keys()
+
+    def __delitem__(self, key):
+        self.s3.rm('/'.join([self.root, key]))
+
+    def __len__(self):
+        return sum(1 for _ in self.keys())
+
+    def __enter__(self):
```
Comment (martindurant) [inline s3fs/mapwrap.py]: I don't have a use case for this.

File: s3fs/mapwrap.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
+
+from collections import MutableMapping
+import os
+
+class MapWrap(MutableMapping):
```
Comment (mrocklin) [inline s3fs/mapwrap.py]: Suggest name `S3Dict` or `S3Map`

File: s3fs/mapwrap.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        if check:
+            s3.touch(root+'/a')
+            s3.rm(root+'/a')
+
+    def clear(self):
+        """Remove all keys below root - empties out mapping
+        """
+        self.s3.rm(self.root, recursive=True)
+
+    def __getitem__(self, key):
```
Comment (mrocklin) [inline s3fs/mapwrap.py]: Suggest 

```python
if isinstance(key, tuple):
    key = list(map(str, key))
else:
    key = str(key)

# in tests
mw[1] = b'hello'
assert mw[1] == b'hello'
del mw[1]
```

Comment (mrocklin) [inline s3fs/mapwrap.py]: In particular I expect to want to use keys like `mw['x', 0, 3]`

**Comment by mrocklin** [comment]
cc @alimanfoo .  S3 MutableMapping coming your way.

File: s3fs/mapwrap.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+import os
+
+class MapWrap(MutableMapping):
+    """Wrap an S3FileSystem as a mutable wrapping.
+
+    The keys of the mapping become files under the given root, and the
+    values (which must be bytes) the contents of those files.
+
+    Usage
+    -----
```
Comment (mrocklin) [inline s3fs/mapwrap.py]: Common header is `Examples`.  Documentation systems will pick up this header and not custom ones like `Usage`.  Also suggest adding a `Parameters` section.

Additionally, for users unfamiliar with the s3fs project we might want to point them to how to build the `s3` object, either by adding a `See Also` link or by including the creation of an `S3FileSystem` object in the example.

File: s3fs/mapwrap.py (hunk 3 above)
Comment (mrocklin) [inline s3fs/mapwrap.py]: Or maybe a general function 

```python
key = normalize_key(key)

def normalize_key(key):
    if isinstance(key, tuple):
        return '/'.join(map(normalize_key, key))
    elif isinstance(key, str):
        return key
    else:
        return key
```

This way we can change around the key naming policy easily in the future if necessary.

**Comment by alimanfoo** [comment]
Awesome!

On Thu, Apr 21, 2016 at 4:12 PM, Matthew Rocklin <notifications@github.com>
wrote:

> cc @alimanfoo <https://github.com/alimanfoo> . S3 MutableMapping coming
> your way.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/dask/s3fs/pull/30#issuecomment-212965350>
>



-- 
Alistair Miles
Head of Epidemiological Informatics
Centre for Genomics and Global Health <http://cggh.org>
The Wellcome Trust Centre for Human Genetics
Roosevelt Drive
Oxford
OX3 7BN
United Kingdom
Email: alimanfoo@googlemail.com <alimanfoo@gmail.com>
Web: http://purl.org/net/aliman
Twitter: https://twitter.com/alimanfoo
Tel: +44 (0)1865 287721

File: s3fs/mapwrap.py (hunk 1 above)
Comment (mrocklin) [inline s3fs/mapwrap.py]: Should probably delete then.

**Comment by mrocklin** [comment]
This looks good to me.  Might want to add it to the API docs.

**Comment by mrocklin** [comment]
Making a few changes.  Will push soon.

File: s3fs/mapping.py [hunk 5]
```diff
@@ -65,24 +63,16 @@ def __setitem__(self, key, value):
             f.write(value)
 
     def keys(self):
-        return iter(map(lambda x: x[len(self.root) + 1:], self.s3.walk(self.root)))
+        return (x[len(self.root) + 1:] for x in self.s3.walk(self.root))
```
Comment (martindurant) [inline s3fs/mapping.py]: The failure must be here: the generator is iterable but not an interator (but gets called by __iter__)

Comment (mrocklin) [inline s3fs/mapping.py]: Generators are Iterators, not Iterables.  

Looking at the travis logs (available if you click the red X) it was an import issue.

```
$ py.test -x --doctest-modules --pyargs s3fs
============================= test session starts ==============================
platform linux -- Python 3.4.4, pytest-2.9.1, py-1.4.31, pluggy-0.3.1
rootdir: /home/travis/build/dask/s3fs, inifile: 
collecting 8 items / 1 errors
==================================== ERRORS ====================================
_________________ ERROR collecting s3fs/tests/test_mapping.py __________________
s3fs/tests/test_mapping.py:2: in <module>
    from s3fs.mapwrap import S3Map
E   ImportError: No module named 's3fs.mapwrap'
!!!!!!!!!!!!!!!!!!!! Interrupted: stopping after 1 failures !!!!!!!!!!!!!!!!!!!!
```

**Comment by martindurant** [comment]
All good from me, if you're done.

**Comment by mrocklin** [comment]
Thanks for the work!  Merged.

**Comment by mrocklin** [comment]
@martindurant is there anything here that would stop this exact same code from working with an `HDFileSystem`?

**Comment by martindurant** [comment]
Probably minor things, like being sure to create the directory first and restrictions on valid key names

**Pull request merged**